Image delete functionality

dodumosu opened the pull request:
this PR implements nditech/apollo-issues#182 and adds front end functionality to delete images.

dodumosu committed:
feat: add image delete permission

also add defaults for roles for that permission.
recreates #894 as a single commit.

see: nditech/apollo-issues#182

dodumosu committed:
chore: add capture attribute on image picker

dodumosu committed:
feat: add image delete functionality

dodumosu committed:
feat: restrict image uploads

only allow image uploads for users with
image deletion permissions

see: nditech/apollo-issues#182

dodumosu committed:
chore: tidy template code

dodumosu committed:
feat: remove image item from gallery

dodumosu committed:
feat: split notifications

takinbo commented on apollo/frontend/templates/frontend/macros/submission_field_macros.html:
```diff
@@ -11,20 +11,24 @@
   <span role="button" tabindex="0"><i class="fas fa-file-upload"></i></span>
   <input type="file" name="{{ form_field.name }}" id="{{ form_field.id }}" data-context="hidden_file" accept=".jpg,.png,image/jpeg,image/png" hidden />
 </label>
+{% endif %}
```
I would have thought the permission only affects the delete button but here it also prevents uploading a replacement. Why?

dodumosu commented:
The linked issue specifically mentioned not allowing uploads as well.

On Wed, 21 Sep 2022, 12:03 Tim Akinbo, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In apollo/frontend/templates/frontend/macros/submission_field_macros.html
> <https://github.com/nditech/apollo/pull/905#discussion_r975978054>:
>
> > +{% if perms.delete_images.can() %}
> +<label class="btn btn-danger mx-2" data-delete-label="{{ form_field.short_name }}"><span role="button" tabindex="1"><i class="fa fa-trash"></i></span></label>
>  <label class="btn btn-secondary mx-2" style="cursor: pointer;" data-context="select_button">
>    <span role="button" tabindex="0"><i class="fas fa-file-upload"></i></span>
>    <input type="file" name="{{ form_field.name }}" id="{{ form_field.id }}" data-context="hidden_file" accept=".jpg,.png,image/jpeg,image/png" hidden />
>  </label>
> +{% endif %}
>
> I would have thought the permission only affects the delete button but
> here it also prevents uploading a replacement. Why?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nditech/apollo/pull/905#pullrequestreview-1114681847>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKX5FROCJ4V6H5MYXMVGWLV7LTP5ANCNFSM6AAAAAAQP7U74M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dodumosu committed:
feat: change permission to modify uploaded images

this commit changes the functionality of the
permission formerly known as the delete images permission.

users without that permission can upload an image where none exists,
but cannot delete or replace existing images

takinbo reviewed (changes requested):
A few changes requested: 

1. The user should be able to delete an image upload that has not been saved. This means that it would not make a backend request to delete the image since it hasn't been uploaded yet.
2. Add the `cursor: pointer` style option when the delete button can be interacted with. Note that simply disabling the button doesn't prevent the pointer from looking the way it's been specified with the style.

dodumosu committed:
feat: clear file input on delete

takinbo commented:
![Screenshot from 2022-10-04 18-19-34](https://user-images.githubusercontent.com/41004/193895656-a71f4ddf-0934-4a83-b579-8905b52915e5.png)

Could you get rid of the notification displayed on the screen after deleting an image.

dodumosu committed:
chore: remove notification toasts

takinbo commented on apollo/frontend/templates/frontend/submission_edit.html:
```diff
@@ -680,9 +701,69 @@ <h5 class="modal-title">{{ _('Add Call Log') }}</h5>
+      imagePresenceLabel.classList.toggle('disabled');
+      imagePresenceLabel.firstChild.firstChild.classList.remove('text-primary');
+      imagePresenceLabel.firstChild.firstChild.classList.add('text-dark');
+      document.querySelector(galleryItemSelector).remove();
```
This assumes that the request will always succeed. Shouldn't it only do this if the response was OK?

dodumosu committed:
fix: add check for response status

takinbo reviewed (changes requested):
Weird behavior: When an image is selected for the first time, the delete button is enabled (correctly) but when another image is selected (without clicking the delete button), it get disabled which shouldn't be.

dodumosu committed:
fix: don't toggle class properties

dodumosu commented:
> Weird behavior: When an image is selected for the first time, the delete button is enabled (correctly) but when another image is selected (without clicking the delete button), it get disabled which shouldn't be.

this should be fixed
